src/detect.c

```diff
@@ -1206,7 +1212,19 @@ static bool DetectRunTxInspectRule(ThreadVars *tv,
                 inspect_flags |= BIT_U32(engine->id);
             }
             break;
+        } else if (!(inspect_flags & BIT_U32(engine->id)) && s->flags & SIG_FLAG_BOTHDIR &&
+                   direction != engine->dir) {
+            // for bidirectional rules, the engines on the opposite direction
```

Comment:
can this be explained more / differently? Not following what is happening.

Reply:
For bidirectional rules, we want to continue and try all the engines we can
`s->app_inspect` are ordered by tx progress, but tx progress is directional.

So you may have a first engine in `s->app_inspect` which is to_client, but you get to inspect the request first.
So, you should skip this engine not in the good direction, but still continue to try other engines

Does it make sense ?

Reply:
I would expect app engines in this list to be ordered by direction and progress, so first toserver 0-N, then toclient 0-N. Is this not how it works?

Reply:
> I would expect app engines in this list to be ordered by direction and progress, so first toserver 0-N, then toclient 0-N. Is this not how it works?

No, it does not work like that, to server and toclient are intertwined by `AppendAppInspectEngine`

And it looks even more complex with the fast_pattern logic that forces one buffer to be first

Reply:
cf SV test rule `alert http any any => any any (msg:"fast_pattern on to_client side"; sid: 7; http.uri; content: "down"; http.server; content: "Apache"; fast_pattern;)` that will first try `http.server`

Debug print in DetectRunTxInspectRule
```diff
+    while (engine != NULL) {
+        printf("lol %d %s\n", s->id, DetectEngineBufferTypeGetNameById(de_ctx, engine->sm_list));
+        engine = engine->next;
+    }
+    engine = s->app_inspect;
```